examples/gno.land/p/demo/nestedpkg/nestedpkg.gno

```diff
@@ -0,0 +1,84 @@
+// Package nestedpkg provides helpers for package-path based access control.
+// It is useful for upgrade patterns relying on namespaces.
+package nestedpkg
+
+import (
+	"std"
+	"strings"
+)
+
+// IsCallerSubPath checks if the caller realm is located in a subfolder of the current realm.
+func IsCallerSubPath() bool {
+	var (
+		cur  = std.CurrentRealm().PkgPath() + "/"
+		prev = std.PrevRealm().PkgPath() + "/"
+	)
+	return strings.HasPrefix(prev, cur)
+}
+
+// AssertCallerIsSubPath panics if IsCallerSubPath returns false.
+func AssertCallerIsSubPath() {
+	var (
+		cur  = std.CurrentRealm().PkgPath() + "/"
+		prev = std.PrevRealm().PkgPath() + "/"
+	)
+	if !strings.HasPrefix(prev, cur) {
+		panic("call restricted to nested packages. current realm is " + cur + ", previous realm is " + prev)
+	}
+}
+
+// IsCallerParentPath checks if the caller realm is located in a parent location of the current realm.
+func IsCallerParentPath() bool {
+	var (
+		cur  = std.CurrentRealm().PkgPath() + "/"
+		prev = std.PrevRealm().PkgPath() + "/"
+	)
+	return strings.HasPrefix(cur, prev)
+}
+
+// AssertCallerIsParentPath panics if IsCallerParentPath returns false.
+func AssertCallerIsParentPath() {
+	var (
+		cur  = std.CurrentRealm().PkgPath() + "/"
+		prev = std.PrevRealm().PkgPath() + "/"
+	)
+	if !strings.HasPrefix(cur, prev) {
+		panic("call restricted to parent packages. current realm is " + cur + ", previous realm is " + prev)
+	}
+}
+
+// IsSameNamespace checks if the caller realm and the current realm are in the same namespace.
+func IsSameNamespace() bool {
+	var (
+		cur  = nsFromPath(std.CurrentRealm().PkgPath()) + "/"
+		prev = nsFromPath(std.PrevRealm().PkgPath()) + "/"
+	)
+	return cur == prev
+}
+
+// AssertIsSameNamespace panics if IsSameNamespace returns false.
+func AssertIsSameNamespace() {
+	var (
+		cur  = nsFromPath(std.CurrentRealm().PkgPath()) + "/"
+		prev = nsFromPath(std.PrevRealm().PkgPath()) + "/"
+	)
+	if cur != prev {
+		panic("call restricted to packages from the same namespace. current realm is " + cur + ", previous realm is " + prev)
+	}
+}
+
+// nsFromPath extracts the namespace from a package path.
+func nsFromPath(pkgpath string) string {
+	parts := strings.Split(pkgpath, "/")
+
+	// Specifically for gno.land, potential paths are in the form of DOMAIN/r/NAMESPACE/...
+	// XXX: Consider extra checks.
+	// XXX: Support non gno.land domains, where p/ and r/ won't be enforced.
+	if len(parts) >= 3 {
+		return parts[2]
```

Comment:
There are already some packages in our examples folder that are nested, so calling `IsSameNamespace` from the nested package and its parent package would return true, even this is not true based on what we consider to be the namespace for upgrade purposes.  

I'd like to understand a bit more what the plan is to solve this problem for package paths of various depths.  And _is_ it really necessary to address this right away?  Maybe not

Reply:
The concept of "nested" "parent" "child" refers to the relationship between an entity (N) and its related entities (N+/-X).

While the concept of sharing a namespace is about sharing the same top-level namespace, more akin to any "family relationship", including brotherhood. The upcoming namespace feature will allow users and teams to own "r/<namespace>", similar to GitHub handles or organization accounts. See #2195.

`IsSameNamespace` is a feature that enables contracts you own, both currently and in the future, to have special powers when they import or call another contract you also own.

Reply:
Got it 👍 . Thanks for the clarification